src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+        return ans;
+      } else {
+        cost.incCurrAns(timeMatrix[currPos][0]);
+        return cost;
+      }
+    }
+
+    /**
+     * Backtracking Step: loop through the time Matrix from currPos, increase
+     * count by 1 and update (int) and (List) value of cost
+     */
+    for (int i = 0; i < numNodes; i++) {
+      // set open boolean and time to add to currenTime
+      boolean isOpen = (boolean) checkIsOpenSetTime(currentTime, i,
+          currPos).get(0);
+      int timeToAdd = (int) checkIsOpenSetTime(currentTime, i,
+          currPos).get(1);
+      
+      // if node is unvisited and greater than 0, i.e. not the same node
+      if (!visited[i] && timeMatrix[currPos][i] > 0 && isOpen) {
+        // mark ith node as visited
+        visited[i] = true;
+
+        // update cost
+        LocalTime timePostUpdate = currentTime.plusMinutes((long) timeToAdd);
+        List<Integer> costPath = cost.getCurrPath();
+        List<Integer> updatePath = new ArrayList<>();
+        updatePath.addAll(costPath);
+        updatePath.add(i);
+        int newTupleCost = cost.getCurrAns() + timeToAdd - HOUR;
+        Tuple updateTuple = new Tuple(newTupleCost, updatePath);
+
+        // recursive call
+        ans = solverHelper(i, numNodes, timePostUpdate, count + 1, 
+            updateTuple, ans, visited);
+
+        // mark ith node as unvisited
+        visited[i] = false;
+      }
+    }
+    return ans;
+  }
+
+  /**
+   * Check if a location is or will be open. Possibilities:
+   * 1. Arrival before open time 
+   * 2. Arrival during open hours
+   * 3. Arrival after closed 
+   * 4. Open Hours are not available
+   * 
+   * @return List<Object> of length 2 
+   *     index 0: (boolean) -- cases 1, 2, 4: true
+   *                           case 3: false
+   *     index 1: (int) --     cases 1, 2, 4: amount of time added by action
+   *                           case 3: Integer.MAX_VALUE
+   * Details per case:
+   * Case 1: add the amount of time till it opens plus time spent (one hour)
+   * Case 2 & 4: add travel time plus time spent (one hour)
+   * Case 3: add Integer.MAX_VALUE
+   */
+  private List<Object> checkIsOpenSetTime(LocalTime currentTime, int i,
+      int currPos) {
+    // list of length 2 to return
+    List<Object> output = new ArrayList<>();
+
+    boolean isOpen; // boolean to be set and added at index 0
+    int timeAdded; // int to be set and added at index 1
+      if (openHours.get(i) != null) {
+        LocalTime open = openHours.get(i).open.time;
+        LocalTime close = openHours.get(i).close.time;
+        
+        // edge case where close time is next day at 0:00
+        if (close.compareTo(LocalTime.parse("00:00")) == 0) {
+          close = LocalTime.of(23, 59);
+        } else {
+          close.minusMinutes(HOUR);
+        }  
+
+        // set time after travelTime
+        LocalTime afterDrive = currentTime.plusMinutes(timeMatrix[currPos][i]);
```

Comment:
I was a little surpized to find this logic here.

I think it's correct, but I'd argue that might be clearer to rename this method somewhat. Eg: rename this method to be something like "isReachableBeforeClosing" or "travelPlusWaitTime", then rename the arguments to be departureTime, departurePoi, targetPoi or similar. That way it's clear that this method is accounting for the travel time.

Reply:
Agreed! The names of method and variables have been changed